### ISSUE TITLE
Should cider work with `--with-poll` emacs?

### ISSUE THREAD

**Issue by jmckitrick** [clojure-emacs/cider#3455]
There's a newish feature in emacs that swaps out `select` for `poll` to allow over 1024 file descriptors to be open, IIUC. It's supposedly a 'drop in' replacement. However, my first attempt to run emacs and cider with this option resulted in a hung connection during 'jack in'. Should this feature even work with cider at all? Or is it worth investigating?

**Comment by jmckitrick** [comment]
I build and run emacs HEAD every week and tinker with settings, so I'm happy to help with this.

**Comment by vemv** [comment]
I'll try out https://github.com/d12frosted/homebrew-emacs-plus which makes this option easy, apparently

(TIL about that distro)

**Comment by jmckitrick** [comment]
@vemv I use that one and https://github.com/daviderestivo/homebrew-emacs-head

**Comment by vemv** [comment]
I ran `brew install emacs-head@30 --with-cocoa --with-poll` and could repro the issue.

By running `M-x toggle-debug-on-error` and then `cider-connect`ing I could see

* First:

```
[nREPL] Establishing direct connection to 127.0.0.1:64844 ...
[nREPL] Direct connection to 127.0.0.1:64844 established
```

* But later:

```
  signal(error ("Sync nREPL request timed out (op clone id 1 time-stamp 2023-09-26 16:32:21.439826000) after 10 secs"))
  error("Sync nREPL request timed out %s after %s secs" ("op" "clone" "id" "1" "time-stamp" "2023-09-26 16:32:21.439826000") 10)
  nrepl-send-sync-request(("op" "clone" "id" "1" "time-stamp" "2023-09-26 16:32:21.439826000") #<buffer *cider-uninitialized-repl*> nil nil)
  nrepl-sync-request:clone(#<buffer *cider-uninitialized-repl*>)
  nrepl--init-client-sessions(#<process nrepl-connection>)
  nrepl-start-client-process("127.0.0.1" 64844 nil #f(compiled-function (_) #<bytecode 0xf69663072ad1ccf>) nil)
  cider-nrepl-connect((:host "127.0.0.1" :port 64844 :project-dir "/Users/vemv/orchard/" :repl-init-function nil :session-name nil :repl-type clj))
```

This would require some investigation.

Perhaps the Emacs 30 changelog explains an adoption guide? i.e. the transition steps.

**Comment by jmckitrick** [comment]
@vemv I saw exactly the same thing. Yet every article I read on 'poll vs select' claimed it was a drop-in replacement that should be transparent to the user. At least, that was my impression.

I'll take a closer look at the emacs docs, but I'm pretty sure there was no migration guide or something similar.

**Comment by jmckitrick** [comment]
@vemv Though I wouldn't be surprised if just a handful of simple changes in cider-nrepl solve the entire issue.

**Comment by vemv** [comment]
Those would be welcome as long that nothing was impacted for the vanilla patch.

**Comment by jmckitrick** [comment]
@vemv The patch is for the macos build... but the feature itself is in the underlying emacs. Let me see if I can find it.

**Comment by jmckitrick** [comment]
But you're right.... this patch is more extensive than I initially thought. Maybe I'll try a vanilla terminal mode emacs and see if that works.

**Comment by jmckitrick** [comment]
@vemv I found the source of the patch!

https://lists.gnu.org/archive/html/emacs-diffs/2022-05/msg00142.html

I was mistaken in thinking this was in emacs master. It's actually on a feature branch. I wonder where the discussion would happen, if this is slated to become part of emacs.

**Comment by vemv** [comment]
Nice!

I can't promise we'll give this a try super soon (other than a quick look and trial/error - already done).

But leaving the issue open seems reasonable, for visibilty. I hear that this capability would be benefitial for LSP performance. OTOH, there's more than one LSP offering for Emacs itself (eglot, lsp-mode), so it wouldn't surprise me if the existent ones are good-enough for the vast majority of cases.

Which is to say, sometimes it's wise to stick to what the authors wanted us to use.

Note that https://lists.gnu.org/archive/html/emacs-diffs/2022-05/msg00142.html is a mailing list item. You can certainly try bumping that conversation or perhaps trying to contact the author.

**Comment by jmckitrick** [comment]
@vemv I emailed him yesterday. I'll report back what I find out.

**Comment by jmckitrick** [comment]
Update: The original author of the feature is not able to work on it at this time, and I'm struggling to get any feedback at all on it.

I'm not sure if this issue is closer to the emacs and poll side, or the cider and nrepl side. I'm open to suggestions on how to proceed. Maybe figuring out where the timeout is actually occurring and if it's just a matter of adjusting how cider/nrepl connects?

**Comment by vemv** [comment]
Just to be clear, is the feature still an unofficial patch?

**Comment by jmckitrick** [comment]
It's an unofficial patch derived from an official feature branch in emacs. I believe the maintainer of `homebrew-emacs-head` chose to extract the code, which was isolated, as a patch rather than have the brew formula build from a separate and possibly stale feature branch.

**Comment by vemv** [comment]
I may be stating the obvious, but if the branch hasn't made it to Emacs stable and its author isn't currently developing the feature, it can be arbitrarily incomplete and buggy?

> I'm not sure if this issue is closer to the emacs and poll side, or the cider and nrepl side. I'm open to suggestions on how to proceed. Maybe figuring out where the timeout is actually occurring and if it's just a matter of adjusting how cider/nrepl connects?

I don't immediately know how to investigate this, but there's a good chance that an investigation would boil down to "upstream feature is buggy".

**Comment by jmckitrick** [comment]
Sigh. I guess we're stuck, then. I just saw a comment saying this would be a really useful feature. That's what got me wondering about it again. Until more people need over 1024 file descriptors in emacs for macOS, it just won't get any more attention.

**Comment by alexander-yakushev** [comment]
I agree that until the feature at least hits the upstream `master`, there is nothing to investigate here really on our side.

**Comment by jakemcc** [comment]
> Sigh. I guess we're stuck, then. I just saw a comment saying this would be a really useful feature. That's what got me wondering about it again. Until more people need over 1024 file descriptors in emacs for macOS, it just won't get any more attention.

Thanks for digging into this, even if the result seems to be a dead end. I'm constantly bumping into this limitation, had heard there was a patch for it but hadn't dug into it, and then saw this issue pop up and got excited that folks were progressing with it.

Looking forward to when this feature lands on master and it is worth digging into it.
